[Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#52949 [Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv**

Enables distributed profiling which we have for gloo and nccl for the MPI backend. Enables it for all collectives as well as send/recv, and also adds shape recording. 

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

rohan-varma committed:
[wip][Dist Profiling] Enable dist profiling for MPI backend

Enables distributed profiling which we have for gloo and nccl for the MPI backend

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][Dist Profiling] Enable dist profiling for MPI backend"


Enables distributed profiling which we have for gloo and nccl for the MPI backend

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][Dist Profiling] Enable dist profiling for MPI backend"


Enables distributed profiling which we have for gloo and nccl for the MPI backend

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][Dist Profiling] Enable dist profiling for MPI backend"


Enables distributed profiling which we have for gloo and nccl for the MPI backend

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[wip][Dist Profiling] Enable dist profiling for MPI backend"


Enables distributed profiling which we have for gloo and nccl for the MPI backend

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv "


Enables distributed profiling which we have for gloo and nccl for the MPI backend. Enables it for all collectives as well as send/recv, and also adds shape recording. 

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv "


Enables distributed profiling which we have for gloo and nccl for the MPI backend. Enables it for all collectives as well as send/recv, and also adds shape recording. 

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv "


Enables distributed profiling which we have for gloo and nccl for the MPI backend. Enables it for all collectives as well as send/recv, and also adds shape recording. 

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv "


Enables distributed profiling which we have for gloo and nccl for the MPI backend. Enables it for all collectives as well as send/recv, and also adds shape recording. 

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv "


Enables distributed profiling which we have for gloo and nccl for the MPI backend. Enables it for all collectives as well as send/recv, and also adds shape recording. 

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv "


Enables distributed profiling which we have for gloo and nccl for the MPI backend. Enables it for all collectives as well as send/recv, and also adds shape recording. 

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv "


Enables distributed profiling which we have for gloo and nccl for the MPI backend. Enables it for all collectives as well as send/recv, and also adds shape recording. 

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma commented:
Change should be ready for review, although still debugging one isend/irecv test @pritamdamania87 @mrzzd

rohan-varma committed:
Update on "[Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv "


Enables distributed profiling which we have for gloo and nccl for the MPI backend. Enables it for all collectives as well as send/recv, and also adds shape recording. 

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv "


Enables distributed profiling which we have for gloo and nccl for the MPI backend. Enables it for all collectives as well as send/recv, and also adds shape recording. 

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

mrzzd commented on torch/lib/c10d/ProcessGroupMPI.hpp:
```diff
@@ -75,13 +75,26 @@ struct WorkEntry {
    public:
-    AsyncWork(at::Tensor tensor, MPI_Request request);
+    AsyncWork(at::Tensor tensor, MPI_Request request,
+    const char * profilingTitle = nullptr,
```
is the indention correct?

rohan-varma replied:
It passed lint, but yeah, I think it should be one level in

mrzzd commented on torch/lib/c10d/ProcessGroupMPI.cpp:
```diff
@@ -144,12 +147,26 @@ int ProcessGroupMPI::AsyncWork::sourceRank() const {
 
 bool ProcessGroupMPI::AsyncWork::wait(std::chrono::milliseconds /* unused */) {
```
does the user has to call wait()? What if for some reason the use does something like this:

while(!work.isComplete()) {  do_more_io_work(); }

rohan-varma replied:
I guess this is technically possible, but if we are designing this similar to RPC API, then we only guarantee event is profiled if user issues a call to `work.wait()` since only after that are we guaranteed that the work has been completed. 

Also, the manual calling of this in `AsyncWork::wait()` is only for send/recv profiling, all other MPI collectives go through the same path as gloo (i.e. call `ProcessGroup::Work::finish`). I guess if user does the following:
```
work = mpi_send(...)
# do a bunch of long-runningother work
work.wait()
```
then the profiling time may be inaccurate, since it would be an overestimate and include the time for the other work in between.

mrzzd replied:
If that's the case, do we have documented this limitation somewhere? I mean the fact that wait is needed for send/recv inorder for profile to work.

rohan-varma replied:
Created an issue mentioning to document this: https://github.com/pytorch/pytorch/issues/53684. We currently don't have much documentation around profiling distributed collectives since the work is landing, but after the main diffs are in, we will add documentation before releasing this feature.

mrzzd reviewed (commented):
Thanks for working in this! My only concern is if wait function is the right place to have finalization logic. Please see my inline comment.

rohan-varma committed:
Update on "[Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv "


Enables distributed profiling which we have for gloo and nccl for the MPI backend. Enables it for all collectives as well as send/recv, and also adds shape recording. 

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]

mrzzd commented:
Looks good to me, thanks! It seems I do not have permission to approve though.

mrzzd reviewed (approved):
Thanks, good to go!

rohan-varma committed:
Update on "[Dist Profiling] Enable dist profiling for MPI backend for collectives and send/recv "


Enables distributed profiling which we have for gloo and nccl for the MPI backend. Enables it for all collectives as well as send/recv, and also adds shape recording. 

Differential Revision: [D26591590](https://our.internmc.facebook.com/intern/diff/D26591590/)

[ghstack-poisoned]
